lib/psql.js

```diff
@@ -24,32 +17,136 @@ function env (db) {
+      process.exit(1)
+    })
+    psql.on('error', (err) => {
+      if (err.code === 'ENOENT') {
+        reject(`The local psql command could not be located. For help installing psql, see https://devcenter.heroku.com/articles/heroku-postgresql#local-setup`)
+      } else {
+        reject(err)
+      }
+    })
+    setTimeout(() => reject('psql call timed out'), timeout)
+  })
+}
+
+function psqlInteractive (dbEnv, prompt, timeout) {
+  const {spawn} = require('child_process')
+  return new Promise((resolve, reject) => {
+    let psql = spawn('psql',
+                     ['--set', `PROMPT1=${prompt}`, '--set', `PROMPT2=${prompt}`],
+                     {env: dbEnv, stdio: 'inherit'})
+    psql.on('error', (err) => {
+      if (err.code === 'ENOENT') {
+        reject(`The local psql command could not be located. For help installing psql, see https://devcenter.heroku.com/articles/heroku-postgresql#local-setup`)
+      } else {
+        reject(err)
+      }
+    })
+    resolve()
+  })
+}
+
+function getConfigs (db) {
+  let dbEnv = env(db)
+  const dbTunnelConfig = tunnelConfig(db)
+  if (db.bastionKey) {
+    dbEnv = Object.assign(dbEnv, {
+      PGPORT: dbTunnelConfig.localPort,
+      PGHOST: dbTunnelConfig.localHost
+    })
+  }
+  return {
+    dbEnv: dbEnv,
+    dbTunnelConfig: dbTunnelConfig
+  }
+}
+
 function handleSignals () {
   process.once('SIGINT', () => {})
 }
 
-function * exec (db, query) {
-  const stripEOF = require('strip-eof')
-  const {spawnSync} = require('child_process')
+function * exec (db, query, timeout = 20000) {
   handleSignals()
-  debug(query)
-  let {stdout, error: err, status} = spawnSync('psql', ['--command', query], {env: env(db), encoding: 'utf8', stdio: [0, 'pipe', 2]})
-  handleError(err)
-  if (status !== 0) process.exit(status)
-  return stripEOF(stdout)
+  let configs = getConfigs(db)
+
+  const tunnelPromise = sshTunnel(db, configs.dbTunnelConfig, timeout)
+  const command = tunnelPromise.then((tnl) => {
+    let psqlPromise = execPsql(query, configs.dbEnv, timeout)
+    return psqlPromise
+  }).catch((error) => {
+    cli.error(error)
+  })
+
+  return yield command
 }
 
 function * interactive (db) {
-  const {spawnSync} = require('child_process')
   const pgUtil = require('./util')
   let name = pgUtil.getUrl(db.attachment.config_vars).replace(/^HEROKU_POSTGRESQL_/, '').replace(/_URL$/, '')
   let prompt = `${db.attachment.app.name}::${name}%R%# `
   handleSignals()
-  let {error: err, status} = spawnSync('psql',
-    ['--set', `PROMPT1=${prompt}`, '--set', `PROMPT2=${prompt}`],
-    {env: env(db), stdio: 'inherit'})
-  handleError(err)
-  if (status !== 0) process.exit(status)
+  let configs = getConfigs(db)
+
+  const tunnelPromise = sshTunnel(db, configs.dbTunnelConfig)
+  const command = tunnelPromise.then((tnl) => {
+    let psqlPromise = psqlInteractive(configs.dbEnv, prompt)
+    return psqlPromise
+  }).catch((error) => {
+    cli.error(error)
+    process.exit(1)
```

Comment:
ideally it should fail with the same status code that `psql` failed with, probably the same in non-interactive